### ISSUE TITLE
Request: relax Cross-Origin-Opener-Policy response header to "same-origin-allow-popups"?

### ISSUE THREAD

**Issue by fasiha** [nolanlawson/pinafore#2261]
**Tldr:** Right now Pinafore is served with `Cross-Origin-Opener-Policy` (COOP) set to `same-origin` and this breaks bookmarklets that use `postMessage`. If this is relaxed to `same-origin-allow-popups`, then when a bookmarklet opens a new window from Pinafore, Pinafore can communicate with the new window via `postMessage`. 

This is demonstrated in this Glitch app: [live site](https://early-fluttering-node.glitch.me) and [code](https://glitch.com/edit/#!/early-fluttering-node). If you open the live site, you can drag and drop the bookmarklet to your bookmark bar. On *most* websites, if you click the bookmarklet, the "popup" page opens, then a second later, the main window `postMessage`s the popup, which receives it, shows an `alert`, and closes the window.

But on Pinafore (and [Buttondown](https://buttondown.email/hillelwayne/archive/please-stop-falling-for-conspiracy-theories/) interestingly), the bookmarklet doesn't work. It opens a new window, but `postMessage`s never arrive. No errors are shown, in macOS Firefox, Chrome, or Safari:

![glitch-popup](https://user-images.githubusercontent.com/37649/204078916-636b2649-8353-4597-bab3-758c2c10bd5f.gif)

The Glitch app itself is served with COOP set to `same-origin` like Pinafore, and you can see the bookmarklet doesn't work for it either. However, by setting the header to `Cross-Origin-Opener-Policy: same-origin-allow-popups`, the bookmarklet *does* work (you'll have to edit the source code to verify, or run locally with HTTPS—sorry).

You probably know a ton more about this header than I do but [this standards document](https://docs.google.com/document/d/1eQEdjHDaJHJ5SFyRKjcqohS7x8XSpF8xmajXimAKcHc/edit#) spells out the difference between `same-origin` and `same-origin-allow-popups` nicely:
> - "same-origin-allow-popups: the document accepts to share its browsing context group only with same-origin documents with COOP policy same-origin-allow-popups AND windows it opens."
> - Same-origin: the document accepts to share its browsing context group only with same-origin documents with COOP same-origin and that have the same CrossOriginEmbedderPolicy (defined in this [draft spec](https://mikewest.github.io/corpp/)).

Between this description and the general observation that other than Pinafore and Buttondown, all other sites work, I wonder if you'd consider relaxing this header to `same-origin-allow-popups`?

Thanks for Pinafore!

**Comment by fasiha** [comment]
(Another example of the same issue: https://github.com/authts/oidc-client-ts/issues/31)

**Comment by nolanlawson** [comment]
Thanks for opening the issue! What is the use case for relaxing this security header? I'm reluctant to increase Pinafore's attack surface unless there is a good use case for it. :slightly_smiling_face:

My main concern is someone tricking someone into clicking a link to Pinafore, which then opens with all of their private data available via `window.localStorage` or `window.indexedDB`, and now the remote site has the ability to do whatever they want, since they have access to the `window` object. At least, that's what I understand relaxing the header would allow for. (IIUC the `same-origin-allow-popups` was mostly designed for OAuth providers / login flows.)

**Comment by fasiha** [comment]
Ahh when you put it that way, I see the value in having this header locked down. My use case is, I have a bookmarking system that I activate with a bookmarklet, and with COOP locked down to `same-site`, Pinafore can’t `postMessage` to the new window it opened. It’s a bit surprising to me that COOP blocks such *outgoing* messages but, like you explain, if it keeps evil sites from accessing Pinafore’s client-side data, then it’s a good trade. I’ll try and adjust my bookmarklet to not use `postMessage`.

Thanks for the super-thoughtful and patient explanation!
